src/bin/cql-stress-cassandra-stress/settings/param/simple_param.rs

```diff
@@ -0,0 +1,159 @@
+use std::{cell::RefCell, fmt, rc::Rc, str::FromStr};
+
+use regex::Regex;
+
+use super::{regex_is_empty, Param, ParamCell, ParamHandle, ParamParseResult};
+
+/// Abstraction of simple parameter which is of the following pattern:
+/// <prefix><value_pattern>
+///
+/// For example `n=` is a simple parameter where
+/// - prefix = "n="
+/// - value_pattern = r"^[0-9]+[bmk]?$"
+pub struct SimpleParam {
+    value: Option<String>,
+    prefix: &'static str,
+    value_pattern: Regex,
+    default: Option<&'static str>,
+    desc: &'static str,
+    required: bool,
+    supplied_by_user: bool,
+    parsed: bool,
+}
+
+impl SimpleParam {
+    pub fn new(
+        prefix: &'static str,
+        value_pattern: &'static str,
+        default: Option<&'static str>,
+        desc: &'static str,
+        required: bool,
+    ) -> Self {
+        Self {
+            value: default.map(|d| d.to_string()),
+            prefix,
+            value_pattern: Regex::new(value_pattern).unwrap(),
+            default,
+            desc,
+            required,
+            supplied_by_user: false,
+            parsed: false,
+        }
+    }
+
+    /// Retrieves the value (if parsed successfully) and consumes the parameter.
+    fn get(self) -> Option<String> {
+        if !self.parsed {
+            return None;
+        }
+        self.value
+    }
+
+    fn is_bool_flag(&self) -> bool {
+        regex_is_empty(&self.value_pattern)
+    }
+}
+
+impl Param for SimpleParam {
+    fn parse(&mut self, arg: &str) -> ParamParseResult {
+        if !arg.starts_with(self.prefix) {
+            return ParamParseResult::NoMatch;
+        }
+
+        if self.supplied_by_user {
+            return ParamParseResult::Error(anyhow::anyhow!(
+                "{} suboption has been specified more than once",
+                self.prefix
+            ));
+        }
+
+        let arg_val = &arg[self.prefix.len()..];
+        if self.value_pattern.is_match(arg_val) {
+            self.supplied_by_user = true;
+            self.value = Some(arg_val.to_string());
+            return ParamParseResult::Success;
+        }
+
+        ParamParseResult::Error(anyhow::anyhow!(
+            "Invalid value {}; must patch pattern {}",
+            arg_val,
+            self.value_pattern.as_str()
+        ))
+    }
+
+    fn supplied_by_user(&self) -> bool {
+        self.supplied_by_user
+    }
+
+    fn required(&self) -> bool {
+        self.required
+    }
+
+    fn set_parsed(&mut self) {
+        self.parsed = true;
+    }
```

Comment:
Why do we need this method at all? It's strange that you first parse the parameter but then you make it appear as if it weren't parsed at all (due to the `self.parsed` flag being false).

I think the `Param` interface should have two methods: one for checking whether given parameter matches to given string, and the other that actually parses it. Only the second method would modify the state of the parameter, so it would simplify reasoning about the state of a `Param`.

Reply:
I agree that this name is not self-explanatory - will change it to something else.

But answering your question - why do we need it?
Assume we declare two parameters:
```
let param1 = parser.simple_param(prefix="param1=", ..., default=None, ...);
// Note that second parameter has some default value.
let param2 = parser.simple_param(prefix="param2=", ..., default=Some("value"), ...);

// Make it so they belong to different groups, i.e. are mutually exclusive.
parser.group(vec![&param1]);
parser.group(vec![&param2]);
```

Now assume that the user provides `param1=something`. This results in the first group being satisfied and their parameters marked with `set_parsed()` (note that param2 is not being marked). Thanks to that, `param2.get()` will return `None` even though it has some default value (meaning, it didn't need to consume any CLI argument). 

So basically what `set_parsed()` does is that it marks the parameters of the group that was satisfied, so the user calling `.get()` method on param handle receives:
- `Some(_)` from the parameters of the satisfied group (if it was provided, or had some default value)
- `None` from the paramteres that doesn't belong to the satisfied group.

Is that clear?

Reply:
I see.

Now that I think of it, how are you going to detect, from the outside, which group has been parsed? By checking which options return Some/None? Perhaps `ParamsParser::parse` should return the index of the matched group? In that case I don't think `set_parsed` should be necessary.

Reply:
> By checking which options return Some/None?
Exactly. The resulting structs hold the `Option`s as well in some cases.
> Perhaps ParamsParser::parse should return the index of the matched group? In that case I don't think set_parsed should be necessary.
That's true, it wouldn't be necessary then. I even thought of it for a moment during implementation. But I am not quite sure how then extraction of the values could look like. Currently, we extract the values via `SimpleParamHandle` API.